// zapmongo/dist/Database.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Database = void 0;
const mongoose_1 = __importDefault(require("mongoose"));
const DatabaseModule_1 = require("./DatabaseModule");
// database manager
class Database {
    constructor(options) {
        this.schemas = new Map();
        // set config
        this.options = options;
        // get loadable schemas
        options.schemas.map((value) => {
            this.schemas.set(value.name, {
                name: value.name,
                data: mongoose_1.default.model(value.name, new mongoose_1.default.Schema(value.data)),
            });
        });
        mongoose_1.default.connect(options.mongoURI, {
            useNewUrlParser: true,
            useUnifiedTopology: true,
        });
    }
    async load(schema) {
        return new DatabaseModule_1.DatabaseModule(this.schemas.get(schema));
    }
}
exports.Database = Database;
